### PR TITLE
Feat: Notifications

### PR DESCRIPTION
**Pull request by MuhammedMagdyy**
##### Why do we need this change?
This PR is for the notifications feature.
- Added new tables called notifications and device token with relation (one-to-many, and one-to-one) to the user, also added an enum for device type with two values only (android and ios).
- Created notifications, deviceToken, and pushNotifications modules.
- Added new endpoints called /notifications, and /device-token.

### CODE REVIEW

File: src/routes/notificationRouter.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        userId,
+      },
+      { pageNumber, pageSize },
+    );
+    if (!notifications) {
+      res.status(404).json({ message: 'Notifications not found' });
+    }
+    res.status(200).json({ notifications });
+  }),
+);
```
Comment (kerolloz) [inline src/routes/notificationRouter.ts:41]: Why are you passing an id here?

File: src/routes/notificationRouter.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      req.body,
+    );
+    const userId = req.user?.id;
+    await notificationService.createOne({
+      ...notification,
+      userId: userId as number,
+    });
+    res.status(201).json({ message: 'Notification created successfully' });
+  }),
+);
```
Comment (kerolloz) [inline src/routes/notificationRouter.ts:21]: Why do we need this?

File: src/routes/notificationRouter.ts (hunk 1 above)
Comment (MuhammedMagdyy) [reply src/routes/notificationRouter.ts:41]: I forgot to delete it.
Fixed ✅

File: src/routes/notificationRouter.ts (hunk 2 above)
Comment (MuhammedMagdyy) [reply src/routes/notificationRouter.ts:21]: I don't know how to create notification, so I added it

File: src/modules/PushNotification/services/pushNotificationService.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+admin.initializeApp({
+  credential: admin.credential.cert(firebasePrivateKey),
+});
+
+export async function sendNotifications(
+  deviceToken: string,
+  title: string,
+  description: string,
+) {
```
[the remark this answers is not in the document]
Comment (MuhammedMagdyy) [reply src/modules/PushNotification/services/pushNotificationService.ts:12]: Fixed ✅

File: src/routes/deviceTokenRouter.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        { id: existingDeviceToken.id },
+        { userId: userId as number },
+      );
+      res.status(200).json({ message: 'Device token updated successfully' });
+    }
+    await deviceTokenService.createOne({
+      ...deviceToken,
+      userId: userId as number,
+    });
+    res.status(201).json({ message: 'Device token created successfully' });
```
Comment (msherifs) [inline src/routes/deviceTokenRouter.ts:42]: hmmm i think this is very complex for an upsert operation, also i recommend to add an upsert function in the deviceToken repo

**Review by msherifs — CHANGES_REQUESTED** [review]

File: src/routes/notificationRouter.ts (hunk 2 above)
Comment (kerolloz) [reply src/routes/notificationRouter.ts:21]: 😅 Okay you can leave it for now as a way to test creating notifications but this one should be removed and you will only send/create notifications when necessary (for example when a user likes or comments on a post, or when there's a new klivvr pick for this week)

Comment (MuhammedMagdyy) [reply src/routes/notificationRouter.ts:21]: Thanks!
I'll add a comment to remind me.

File: src/modules/PushNotification/services/pushNotificationService.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        notification: {
+          title,
+          body: description,
+        },
+        token: deviceToken,
+      };
+      await admin.messaging().send(message);
+    } catch (err) {
+      throw new Error();
+    }
```
Comment (kerolloz) [inline src/modules/PushNotification/services/pushNotificationService.ts:35]: What's the point of a try-and-catch here when you're throwing an error eventually ?

**Review by kerolloz — APPROVED** [review]

**Review by msherifs — DISMISSED** [review]

**Review by kerolloz — APPROVED** [review]

**Review by msherifs — APPROVED** [review]

**Review by mohamadelhinamy — APPROVED** [review]

**Pull request merged**